src/dragonball/src/resource_manager.rs

```diff
@@ -0,0 +1,683 @@
+            msi_irq_pool: Mutex::new(self.msi_irq_pool),
+            pio_pool: Mutex::new(self.pio_pool),
+            mmio_pool: Mutex::new(self.mmio_pool),
+            mem_pool: Mutex::new(self.mem_pool),
+            kvm_mem_slot_pool: Mutex::new(self.kvm_mem_slot_pool),
+        }
+    }
+}
+
+/// Resource manager manages all resources for a virtual machine instance.
+pub struct ResourceManager {
+    legacy_irq_pool: Mutex<IntervalTree<()>>,
+    msi_irq_pool: Mutex<IntervalTree<()>>,
+    pio_pool: Mutex<IntervalTree<()>>,
+    mmio_pool: Mutex<IntervalTree<()>>,
+    mem_pool: Mutex<IntervalTree<()>>,
+    kvm_mem_slot_pool: Mutex<IntervalTree<()>>,
+}
+
+impl Default for ResourceManager {
+    fn default() -> Self {
+        ResoureManagerBuilder::default().build()
+    }
+}
+
+impl ResourceManager {
+    /// Create a resource manager instance.
+    pub fn new(max_kvm_mem_slot: Option<usize>) -> Self {
+        let res_manager_builder = ResoureManagerBuilder::default();
+        res_manager_builder
+            .init_legacy_irq_pool()
+            .init_msi_irq_pool()
+            .init_pio_pool()
+            .init_mmio_pool()
+            .init_mem_pool()
+            .init_kvm_mem_slot_pool(max_kvm_mem_slot)
+            .build()
+    }
+
+    /// Init mem_pool with arch specific constants.
+    pub fn init_mem_pool(&self) {
+        let mut mem = self.mem_pool.lock().unwrap();
+        ResoureManagerBuilder::init_mem_pool_helper(&mut mem);
+    }
+
+    /// Check if mem_pool is empty.
+    pub fn is_mem_pool_empty(&self) -> bool {
+        self.mem_pool.lock().unwrap().is_empty()
+    }
+
+    /// Allocate one legacy irq number.
+    ///
+    /// Allocate the specified irq number if `fixed` contains an irq number.
+    pub fn allocate_legacy_irq(&self, shared: bool, fixed: Option<u32>) -> Option<u32> {
+        // if shared_irq is used, just return the shared irq num.
+        if shared {
+            return Some(SHARED_IRQ);
+        }
+
+        let mut constraint = Constraint::new(1u32);
+        if let Some(v) = fixed {
+            if v == SHARED_IRQ {
+                return None;
+            }
+
+            constraint.min = v as u64;
+            constraint.max = v as u64;
+        }
+        // Safe to unwrap() because we don't expect poisoned lock here.
+        let mut legacy_irq_pool = self.legacy_irq_pool.lock().unwrap();
+        let key = legacy_irq_pool.allocate(&constraint);
+        if let Some(k) = key.as_ref() {
+            legacy_irq_pool.update(k, ());
+        }
+        key.map(|v| v.min as u32)
+    }
+
+    /// Free a legacy irq number.
+    ///
+    /// Panic if the irq number is invalid.
+    pub fn free_legacy_irq(&self, irq: u32) {
+        // if the irq number is shared_irq, we don't need to do anything.
+        if irq == SHARED_IRQ {
+            return;
+        }
+
+        if !(LEGACY_IRQ_BASE..=LEGACY_IRQ_MAX).contains(&irq) {
+            panic!("invalid irq number when freeing legacy irq");
```

Comment:
If we really have to panic, please add a unit test using `#[should_panic]` for this scenario.

Reply:
The panic has been removed.

We use `Result` to return errors.